### PR TITLE
Add nuphar python scripts to wheel, and notebook tutorial

### PR DESCRIPTION
**Pull request by KeDengMS**
**Description**:This change adds tutorial of Nuphar in Python notebook

**Motivation and Context**
- Helps user to understand Nuphar better

### CODE REVIEW

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+   "metadata": {},
+   "source": [
+    "# ONNX Runtime: Tutorial for Nuphar execution provider\n",
+    "**Accelerating model inference via compiler, using Docker Images for ONNX Runtime with Nuphar**\n",
+    "\n",
+    "This example shows how to accelerate model inference using Nuphar.\n",
+    "\n",
+    "#### Tutorial Roadmap:\n",
+    "1. Create and run inference on a simple ONNX model, and understand how ***compilation*** works in Nuphar.\n",
+    "2. Create and run inference on a ***LSTM*** model, run symbolic shape inference, edit LSTM ops to Scan, and check Nuphar speedup.\n",
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: 'a LSTM' -> 'an LSTM'

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+    "onnx.save(model, simple_model_name)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When loading the model we created above, Nuphar execution provider would be used. We can use settings string to check the generated code.\n",
+    "\n",
+    "Because of redirection of output, we dump the lowered code from a subprocess to a log file:"
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: "redirection" -> "the redirection"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+    "# save the model\n",
+    "simple_model_name = 'simple.onnx'\n",
+    "onnx.save(model, simple_model_name)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When loading the model we created above, Nuphar execution provider would be used. We can use settings string to check the generated code.\n",
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: Probably modify the sentence to "We will use nuphar execution provider to run the inference for the model that we created above"?

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+    "    log_lines = f.readlines()\n",
+    "\n",
+    "log_lines[-10:]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The compiled code shows that the nodes of Add/Mul is fused in a single function, and vectorization is applied in the loop. This fusion happened inside compiler used by Nuphar execution provider, and it does not require model editing.\n",
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: "is fused in" -> "are fused into"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+    "\n",
+    "Next, let's run inference on the model and compare with numpy in accuracy and speed:"
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: probably modify the sentence to "and compare the accuracy and performance with numpy"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+    "'onnxruntime: {} seconds, numpy: {} seconds'.format(end_ort - start_ort, end_np - start_np)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "## 2. Create and run inference on a LSTM model\n",
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: "a LSTM" -> "an LSTM"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+    "generate_model('lstm', input_dim, hidden_dim, bidirectional=False, layers=4, model_name=lstm_model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "**IMPORTANT**: since compiler generates code before data is fed, shape inference information is critical for optimization in compiler. To do that we run symbolic shape inference on the model."
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typos: 
"compiler" -> "the compiler"

"data" -> "the input data"

"optimization in compiler" -> "compiler optimizations"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
+    "SymbolicShapeInference.infer_shapes(input_model=lstm_model, output_model=lstm_model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "Now, let's check basline performance on the generated model, using CPU execution provider."
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: basline -> the baseline

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 9]
```diff
@@ -0,0 +1,681 @@
...
+    "input_data = np.random.rand(seq, 1, input_dim).astype(np.float32)\n",
+    "feed = {sess_baseline.get_inputs()[0].name:input_data}\n",
+    "output = sess_baseline.run([], feed)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To run RNN models in Nuphar execution provider efficiently, user need to first convert LSTM/GRU/RNN ops to Scan ops. This is because Scan is more flexible and supports quantized RNNs."
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: "user need" -> "the user needs"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 10]
```diff
@@ -0,0 +1,681 @@
...
+    "\n",
+    "'nuphar: {} seconds, baseline: {} seconds'.format(end_nuphar - start_nuphar, end_baseline - start_baseline)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Quantize LSTM model\n",
+    "Let's get more speed-ups from Nuphar by quantizing the floating point GEMM/GEMV in LSTM model to int8 GEMM/GEMV.\n",
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: "LSTM model" -> "the LSTM model"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 11]
```diff
@@ -0,0 +1,681 @@
...
+    "from onnxruntime.nuphar.model_quantizer import convert_matmul_model\n",
+    "quantized_model = 'Scan_LSTMx4_int8.onnx'\n",
+    "convert_matmul_model(scan_model, quantized_model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now run the quantized model, and check accuracy. Please note that quantization brings some error, so we relaxed on compare threshold."
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: "quantization brings some error, so..." -> "quantization may cause losses of accuracy, so we relax the comparison threshold a bit."?

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 12]
```diff
@@ -0,0 +1,681 @@
...
+    "\n",
+    "'quantized: {} seconds, non-quantized: {} seconds'.format(end_quantized - start_quantized, end_nuphar - start_nuphar)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 4. Working on a real model: Bidirectional Attention Flow (BiDAF)\n",
+    "BiDAF is a machine comprehension model that uses LSTM. The inputs to this model are paragraphs of context and query, and the outputs are start/end index of words in the context that answers the query.\n",
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: 
"of context and query, and the outputs are start/end index of words in the context that answers the query" ->

"of contexts and queries, and the outputs are start/end indexes of the words in the contexts that answer the queries"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 13]
```diff
@@ -0,0 +1,681 @@
...
+    "  urllib.request.urlretrieve(bidaf_url, bidaf_local)\n",
+    "with tarfile.open(bidaf_local, 'r') as f:\n",
+    "  f.extractall(cwd)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's check the performance in CPU provider:"
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: "the performance in CPU provider" -> "the performance of the CPU provider"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb [hunk 14]
```diff
@@ -0,0 +1,681 @@
...
+    "from onnxruntime.nuphar.symbolic_shape_infer import SymbolicShapeInference\n",
+    "from onnxruntime.nuphar.model_editor import convert_to_scan_model\n",
+    "from onnxruntime.nuphar.model_quantizer import convert_matmul_model\n",
+    "\n",
+    "# editing\n",
+    "bidaf_converted = 'bidaf_mod.onnx'\n",
+    "SymbolicShapeInference.infer_shapes(bidaf, bidaf_converted)\n",
+    "convert_to_scan_model(bidaf_converted, bidaf_converted)\n",
+    "# When quantizing, there's an only_for_scan option to quantize only the GEMV inside Scan ops.\n",
+    "# This is useful when the input dims of LSTM is much bigger than hidden dims.\n",
```
Comment (yangchen-MS) [inline docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: typo: "...dims of LSTM is" -> "dims of LSTM are"

File: docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb (hunk 1 above)
Comment (KeDengMS) [reply docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: L is not vowel?

Comment (yangchen-MS) [reply docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb]: For an acronym, it's based on the pronunciation of the first letter. In this case, it's "eL", where 'e' is  a vowel.

**Pull request closed without merging**